pkgs/by-name/po/polymc-unwrapped/package.nix

```
@@ -0,0 +1,103 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+  cmark,
+  extra-cmake-modules,
+  gamemode,
+  ghc_filesystem,
+  jdk8,
+  kdePackages,
+  ninja,
+  nix-update-script,
+  stripJavaArchivesHook,
+  tomlplusplus,
+  zlib,
+
+  msaClientID ? null,
+  gamemodeSupport ? stdenv.hostPlatform.isLinux,
+}:
+
+let
+  libnbtplusplus = fetchFromGitHub {
+    owner = "PolyMC";
+    repo = "libnbtplusplus";
+    rev = "2203af7eeb48c45398139b583615134efd8d407f";
+    hash = "sha256-TvVOjkUobYJD9itQYueELJX3wmecvEdCbJ0FinW2mL4=";
+  };
+in
+
+assert lib.assertMsg (
+  gamemodeSupport -> stdenv.hostPlatform.isLinux
+) "gamemodeSupport is only available on Linux.";
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "polymc-unwrapped";
+  version = "7.0";
+
+  src = fetchFromGitHub {
+    owner = "PolyMC";
+    repo = "PolyMC";
+    rev = "refs/tags/${finalAttrs.version}";
+    hash = "sha256-d6Npu21iC5JUhWOM2DMn7gtzOj/Oo+ZoB36birHxzyA=";
+  };
+
+  postUnpack = ''
+    rm -rf source/libraries/libnbtplusplus
+    ln -s ${libnbtplusplus} source/libraries/libnbtplusplus
+  '';
+
+  nativeBuildInputs = [
+    cmake
+    ninja
+    extra-cmake-modules
+    jdk8
+    stripJavaArchivesHook
+  ];
+
+  buildInputs = [
+    cmark
+    ghc_filesystem
+    kdePackages.qtbase
+    kdePackages.qtcharts
+    kdePackages.quazip
+    tomlplusplus
+    zlib
+  ] ++ lib.optional gamemodeSupport gamemode.dev;
+
+  hardeningEnable = lib.optionals stdenv.hostPlatform.isLinux [ "pie" ];
+
+  cmakeFlags =
+    [
+      (lib.cmakeFeature "Launcher_BUILD_PLATFORM" "nixpkgs")
+      (lib.cmakeFeature "Launcher_QT_VERSION_MAJOR" (lib.versions.major kdePackages.qtbase.version))
```

```suggestion
      (lib.cmakeFeature "Launcher_QT_VERSION_MAJOR" (lib.versions.major kdePackages.qtbase.version))
      (lib.cmakeBool "Launcher_STRICT_DRM" true)
```
We'll probably want to enable this for obvious reasons

Without it, the launcher will allow users to download and play the game without ever verifying if they actually own it. This has been [discussed before](https://github.com/NixOS/nixpkgs/pull/313011#pullrequestreview-2065269956) in regards to other MultiMC-related forks, is in pretty clear violation in Mojang's TOS, and may actually be illegal in some jurisdictions

This launcher may also do things to break the TOS of other services (like Curseforge, as it mentions after the setup wizard), but it seems to properly inform people and isn't really stepping into the territory of legal troubles; this feature does not do that and AFAICT the build option is the only way to toggle it

---

I'm fine with adding this if this is absolutely necessary, but I would prefer not to. Mostly based on the premises discussed in the original issue #365781. In the final decision of the steering committee:

> We, the Nix Steering Committee, believe that Nixpkgs can accept the contribution of a PolyMC package, so long as whoever maintains that package abides by our Code of Conduct and values and there are no other reasons for rejection.

I can find no reference of this in the Code of Conduct to DRM circumvention being disallowed as it primarily focuses on interpersonal behavior in the community (which I hope and believe I have followed the standards set forth in).

Further, I feel as though DRM violating software and questionable legality (albeit only in specific jurisdictions) has been something that Nixpkgs has consistently stepped beyond (see yt-dlp, MakeMKV, libdvdcss/libaacs/libbdplus, and pretty much anything having to do with physical media or ebook collection). At the end of the day, there is no special thing going on in the background for this but a simple if statement requiring a premium account in the launcher. The content of the game is downloaded the same, and there is no trickery or attempt to conceal the activity from Mojang.

I'd rather package this as closely to what is intended by the team behind it, in the words of Lenny:

> [The team] believe[s] in user freedom, it makes no sense to impose an arbitrary DRM restriction (which generally hurt legitimate users more than anyone else) in a libre software project. The only reason it's even kept as a compile time option is in order to comply with some distribution platforms potentially requiring it to be set.

---

Given all of the above and that I don’t see any CoC or policy conflict with adding PolyMC, I support the PR landing as is.

---

> In the final decision of the steering committee:
> 
> > [...]
> 
> I can find no reference of this in the Code of Conduct to DRM circumvention being disallowed as it primarily focuses on interpersonal behavior in the community (which I hope and believe I have followed the standards set forth in).

I never meant to imply that this in violation of the Code of Conduct. Like you said, the CoC primarily focuses on interpersonal behavior; this is a technical and legal issue

> We, the Nix Steering Committee, believe that Nixpkgs can accept the contribution of a PolyMC package, so long as whoever maintains that package abides by our Code of Conduct and values and there are no other reasons for rejection.

I don't believe this statement is very relevant to this topic in the first place, though. Much of the original thread, and this entire response is more-so focused on the behavior of upstream, differentiating contributors like yourself from them, and possible security concerns. In fact, the only mention (I can find) of this flag allowing you to essentially pirate the game is [here](https://github.com/NixOS/nixpkgs/issues/365781#issuecomment-2552199580), and it was only really acknowledged by an upstream developer [here](https://github.com/NixOS/nixpkgs/issues/365781#issuecomment-2552206483)

> [The team] believe[s] in user freedom, it makes no sense to impose an arbitrary DRM restriction (which generally hurt legitimate users more than anyone else) in a libre software project. The only reason it's even kept as a compile time option is in order to comply with some distribution platforms potentially requiring it to be set.

And as much as we should appreciate user freedom, it's important to note how implementing (or removing) features like this can impact the project at large. The distributions that require this flag to be set usually do so in order to protect themselves from potential legal trouble, which could very well endanger them and undermine that same goal of user freedom

-----

So to hopefully clear things up, I want to ask @NixOS/steering:

- What is the committee's stance on packages that may **directly violate agreements with third parties**?
  - i.e., terms of service, end user license agreements
  - Packages with specific code to bypass, workaround, or otherwise ignore said agreements
- What is the committee's stance on packages that may **directly enable piracy**?
  - Specifically, packages with code meant to work around protections. Not packages that could be use indirectly like a (de)compiler, torrent client, or web browser
  - Examples may include direct game "cracks", specific client modifications, etc.
- Should we alert users of these possibilities?
  - Some packages (like this) may cause users to unknowingly violate agreements with third party vendors or even enter legal grey areas
  - This is very inconsistent currently, with some examples being:
    - The [Minecraft NixOS module](https://github.com/NixOS/nixpkgs/blob/d2f6f28bc5c823e1817d68769f3c8e389b0b41f0/nixos/modules/services/games/minecraft-server.nix#L318-L324) requiring users to agree to the EULA
    - [Bluemap](https://github.com/NixOS/nixpkgs/blob/d2f6f28bc5c823e1817d68769f3c8e389b0b41f0/nixos/modules/services/web-apps/bluemap.nix#L309-L315) doing the same thing
    - yt-dlp being packaged with no warning, as mentioned above
    - Some packages (like [`corefonts`](https://github.com/NixOS/nixpkgs/blob/d2f6f28bc5c823e1817d68769f3c8e389b0b41f0/pkgs/by-name/co/corefonts/package.nix#L127-L128)) installing the EULA, but not giving any notice or ensuring it's been agreed to
- How should these packages be handled in regards to foundation resources like the hydra.nixos.org cache?
  - Obligatory IANAL, but AFAIK the Foundation could face legal trouble for hosting binaries for some packages through laws like the DMCA
  - Unfree packages are restricted from building for similar reasons
  - Copyright-related actions taken against open source hosters seem to be getting more common, with Nintendo's DMCA notices to GitHub being a recent example
  
Thanks in advance :)

---

At least on the legal end, while also not being a lawyer, I do have a fair enough background to provide some thought on this.

> [T]he Foundation could face legal trouble for hosting binaries for some packages through laws like the DMCA

Possible, but in this case pretty remote. There’s no "circumvention device" under 17 U.S.C. §1201 here; the launcher merely skips an account check, and Mojang itself serves the client archives over plain HTTP with zero authentication. The situations that actually slip into a gray zone are those where explicit reverse‑engineering comes into question (Switch emulators, AACS/CSS/BD+/ADE strippers, etc.). I’m not aware of any precedent where downstream packagers were hit for hosting anything of this nature.

So the real question is whether the cache ships anything that isn’t free and open source. And that isn't the case here, we’re just providing a launcher that pulls those files onto the user’s machine. I don’t see any DMCA problem for the Foundation.

On the notice point: flagging every package that might tick off a vendor would be a nightmare. By that logic we’d slap warnings on half of nixpkgs, think video players bundling `libdvdcss`, Calibre, and countless others. And no mainstream package manager does that. Since free software has never asked a conglomerate’s permission to hit its unsecured endpoints, I don’t think nixpkgs should start now.

---

Here's an idea: Make the DRM CMake flag an overridable boolean input (on by default).

Therefore, you're not giving anyone dirt to throw at you. I'm pretty sure 90% of NixOS users already have minecraft bought anyway, and those who don't will either ask or look at the package source for a solution.


In almost all jurisdictions it's perfectly legal to have pirated copies, but not distribute them. This technicality is what makes torrenting something like S.T.A.L.K.E.R 2 illegal but not downloading the same pirated copy from a filehost (again, due to technicalities in the Torrent protocol vs a download over http or whatever).

Last thing you want is another round of drama like the divide with Anduril's sponsorship.

---

> Here's an idea: Make the DRM CMake flag an overridable boolean input (on by default).

This is the outcome I'd like to see as well, similar to unfree packages. As long as the user is informed, opt-ing in, and sources/binaries aren't being hosted by the Foundation, options like this are pretty harmless IMHO

---

I'd like to see the some official stance on this before I'd make such a change. Steering has been mentioned so we can wait. I really have no problem with adding that, but it just feels like an unnecessary compromise especially given 1) that Lenny says this is not how the launcher is intended to be and 2) the precedent set by having packages all over the place which would violate Terms of Services, EULAs, etc. I reiterate yt-dlp as an example of this.

---

How does yt-dlp violate TOS and EULA? All it does is download video content from a webpage

---

In the case of yt-dlp, the user is breaking the terms of service, not the software. There's no clause mentioning about distributing and/or creating software that enables you to break their terms of service, because they'd mean that every single browser would be forbidden (and thus you can't use the service).

However, it's much easier to write a legal document that forbids unauthorized distribution of software copies, which Minecraft could very well also be subject to such clause(s).


Just make it depend of an input (for example, call it hasDRM), and when you include your package (and want DRM disabled) you'd just do `pkgs.polymc.override { hasDRM = false; }`

---

And FYI, this is how people went around laws in the US in the early 90's prohibiting encryption.

Can't distribute GPG because the NSA can't spy on everyone? Cool. Just distribute the source code and have it be protected under "free speech".

This is what we're going to do here.

---

> In the case of yt-dlp, the user is breaking the terms of service, not the software.

I'm not sure how you mean. With `yt-dlp`, you give it a YouTube link and effectively say "I want this video." In the backend, it has a hack explicitly written for the purpose of extracting video segments from `youtube.com`. It is a cat and mouse game, an older version of `yt-dlp` won't work for you because it is constantly being patched and fixed.

So in that case, the software is actively finding ways to circumvent DRM. YouTube disallows downloading their content and actively pursues methods to stop it from happening. On the other hand, you have this, where all it does is remove an `if` statement checking if you have a valid account connected. Keep in mind, the actual game has already been downloaded and installed on the system. There is no access control circumvention like in `yt-dlp`.

---

Also: I took the time to read Minecraft's EULA and I actually don't think it can be argued that the software without the check would break this license agreement. Below are a few excerpts that *could* be related. I don't think any of these are violated by this software.

> We need all game downloads and updates to come from a source that we authorize.

> However, you must not distribute anything we've made unless we specifically agree to it.

> By 'distribute anything we've made' what we mean is: give copies of our game software or content to anyone else; make commercial use of anything we've made; try to make money from anything we've made; or let other people get access to anything we've made in a way that is unfair or unreasonable.

This last one is the closest I can think of that would apply, but even then it is vague. Specifically, it disallows "let[ting] other people get access to anything we've made in a way that is unfair or unreasonable." It is my opinion that distributing software that taps into unsecured endpoints IS NOT unfair or unreasonable. If the Foundation is still hesitant regarding this, I know a lawyer well and can get a quick opinion on this.

We are not:

- distributing Minecraft;
- modifying or redistributing a Modded Version; or
- hacking or breaching security.

---

<blockquote>

I am active and eligible to practice in Texas under the State Bar. I provide what I know with confidence that it is correct to my understanding of intellectual property law, but I guarantee nothing and do not wish to create an attorney-client relationship with the so-called Foundation.

From what I've gathered, the software in question does not itself distribute the contents of the video game and does not obtain the contents of the video game in any questionable manner - as in, it uses the official distribution internet service to down load the video game. At its face, this doesn't seem to be problematic as it is well established in intellectual property law that in order to enforce your rights over a specific property, you must make some semblance of an attempt to protect the property. Think of it in the context of real land ownership, if one does not have a gate or signage warning of private property, you cannot immediately pursue criminal trespass.

So the question then becomes whether or not the official method of access to the video game (i.e. Microsoft's launcher) containing an account check is classified as an access control measure under the DMCA, and if using an alternative access system that fails to include this is therefore a circumvention of that technological measure. There is very little precedent in the courts for this and it is challenging to confidently classify this action using the text within the DMCA, but I think there might be just enough to make a reasonable assertion here.

Critically, courts have been reluctant to stretch the scope of "access controls" to include mechanisms that are primarily procedural or contractual, rather than technological in function. In this case, the Microsoft launcher may be part of Microsoft's user management infrastructure, but the video game files themselves are neither encrypted nor technically restricted beyond that procedural check. Once lawfully downloaded using authorized channels, the user possesses the full contents of the game and can likely use them as he wishes.

Moreover, with respect to the EULA: the enforceability of such agreements relies heavily on when and how the user assents to the terms. If the alternative launcher accesses files already lawfully obtained, and the user has not affirmatively accepted terms that could prohibit this method of execution, then any claim of breach becomes tenuous. Even assuming, arguendo, that the EULA was assented to, a violation of contract does not equate to copyright infringement, as affirmed by the Ninth Circuit in MDY v. Blizzard (9th Cir. 2010). There, the court drew a clear line between contractual terms and the rights under copyright, holding that breaching a EULA only implicates copyright law when it exceeds the scope of the license AND infringes on an exclusive right.

This suggests that there is a small legal gray area for the user of the alternative launcher, but I believe this is the only legal limitation that can be discussed in relation to the issue at hand. To answer the core question of if the operator of any software distributor needs to fear consequence for distributing this launcher: I would venture to say your exposure is close to zero.
</blockquote>

---

> Also: I took the time to read Minecraft's EULA and I actually don't think it can be argued that the software without the check would break this license agreement. Below are a few excerpts that _could_ be related. I don't think any of these are violated by this software.
> 
> > We need all game downloads and updates to come from a source that we authorize.
> 
> > However, you must not distribute anything we've made unless we specifically agree to it.
> 
> > By 'distribute anything we've made' what we mean is: give copies of our game software or content to anyone else; make commercial use of anything we've made; try to make money from anything we've made; or let other people get access to anything we've made in a way that is unfair or unreasonable.
> 
> This last one is the closest I can think of that would apply, but even then it is vague. Specifically, it disallows "let[ting] other people get access to anything we've made in a way that is unfair or unreasonable." It is my opinion that distributing software that taps into unsecured endpoints IS NOT unfair or unreasonable. If the Foundation is still hesitant regarding this, I know a lawyer well and can get a quick opinion on this.
> 
> We are not:
> 
>     * distributing Minecraft;
> 
>     * modifying or redistributing a Modded Version; or
> 
>     * hacking or breaching security.

Even if you're right, this could potentially turn us into a big legal target (especially if Microsoft feels like getting a couple million. I assume you know how broken the US legal system is).

Then, everyone would be arguing about "muh heckin morals" or "muh heckin microsoft will suck our funding dry!"


Just add the toggle. Later you can try setting the DRM to be disabled without anyone noticing in a future PR.